core/src/main/java/com/linecorp/armeria/common/stream/FixedStreamMessage.java

```diff
@@ -0,0 +1,309 @@
+        final SubscriptionImpl subscription = this.subscription;
+        if (subscription == null) {
+            return;
+        }
+
+        if (fulfilled == requested) {
+            return;
+        }
+
+        if (subscription.needsDirectInvocation()) {
+            notifySubscriber0(subscription);
+        } else {
+            subscription.executor().execute(() -> notifySubscriber0(subscription));
+        }
+    }
+
+    private void notifySubscriber0(SubscriptionImpl subscription) {
+        if (inOnNext) {
+            // Do not let Subscriber.onNext() reenter, because it can lead to weird-looking event ordering
+            // for a Subscriber implemented like the following:
+            //
+            //   public void onNext(Object e) {
+            //       subscription.request(1);
+            //       ... Handle 'e' ...
+            //   }
+            //
+            // Note that we do not call this method again, because we are already in the notification loop
+            // and it will consume the element we've just added in addObjectOrEvent() from the queue as
+            // expected.
+            //
+            // We do not need to worry about synchronizing the access to 'inOnNext' because the subscriber
+            // methods must be on the same thread, or synchronized, according to Reactive Streams spec.
+            return;
+        }
+
+        if (!invokedOnSubscribe) {
+            // Subscriber.onSubscribe() was not invoked yet.
+            // Reschedule the notification so that onSubscribe() is invoked before other events.
+            //
+            // Note:
+            // The rescheduling will occur at most once because the invocation of onSubscribe() must have been
+            // scheduled already by subscribe(), given that this.subscription is not null at this point and
+            // subscribe() is the only place that sets this.subscription.
+
+            subscription.executor().execute(() -> this.notifySubscriber0(subscription));
+            return;
+        }
+
+        final Subscriber<Object> subscriber = subscription.subscriber();
+        for (;;) {
+            if (closeEvent != null) {
+                cleanup();
+                return;
+            }
+
+            if (fulfilled == objs.length) {
+                notifySubscriberOfCloseEvent(subscription, SUCCESSFUL_CLOSE);
+                return;
+            }
+
+            final long requested = this.requested;
+
+            if (fulfilled == requested) {
+                break;
+            }
+
+            while (fulfilled < requested) {
+                if (closeEvent != null) {
+                    cleanup();
+                    return;
+                }
+
+                T o = objs[fulfilled];
+                objs[fulfilled++] = null;
+                o = prepareObjectForNotification(subscription, o);
+                inOnNext = true;
+                try {
+                    subscriber.onNext(o);
```

Comment:
Just curious. Do we need to release `o` if an exception is raised from `onNext`? Because `fulfilled` was already increased, so `o` would be missed to be released in `cleanup`.

Reply:
`prepareObjectForNotification` releases when subscriber is not aware of pooled objects. I think when it is aware, it's pretty clear that ownership is immediately transfered by the call to `onNext`, and it's up to the subscriber to release at that point even if it throws an exception. All of our streams implement this behavior.

Reply:
Ok, thanks! :-)